snuba/request/request_settings.py

```diff
@@ -1,10 +1,69 @@
+from __future__ import annotations
+
 from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
 
-from typing import Sequence
+from typing import Any, Mapping, Optional, Sequence
 
 from snuba.state.rate_limit import get_global_rate_limit_params, RateLimitParameters
 
 
+class SamplingMode(Enum):
+    """
+    The value of this ennum are part of the public schema. Do not change them
+    without ensuring backward compatibility.
+    """
+
+    # No sampling applied.
+    NONE = "none"
+    # A fixed 0 - 1.0 (both exclusive) sampling rate is provided.
+    FIXED = "fixed"
+    # Adaptive sampling rate with aggregation adjustment. In this mode Snuba
+    # tries to guess an appropriate sampling rate depending on several heuristics
+    # including stats on the expected amount of data to query.
+    # Snuba will try its best to adjust the result of aggregations according to
+    # the sampling rate where it makes sense. If an aggregation function is passed
+    # that Snuba does not know how to adjust, Snuba will bail and apply no sampling
+    # so the client will still receive data that does not need adjustment, thus
+    # copmliant with what was requested.
+    ADAPTIVE_ADJUST = "adaptive_adjust"
+    # Adaptive sampling rate without aggregation adjustment. It works like for
+    # adaptive_adjust but Snuba will not apply any adjustment. It is responsibility
+    # of the client to apply the appropriate adjustment.
+    # This gives higher guarantees that sampling will actually be applied since
+    # Snuba does not need to know how to adjust the aggregations provided in the
+    # query.
+    ADAPTIVE_NO_ADJUST = "adaptive_no_adjust"
```

Comment:
extreme nit: `ADAPTIVE_ADJUSTED` and `ADAPTIVE_UNADJUSTED` seem like better names to me.

Also, similar to [this comment](https://github.com/getsentry/snuba/pull/802/files#r386737896), the sampling mode seems like something that would be important to include in the response and query log.

Reply:
I would be happy to add it to the query log/debug, but I was actually deliberate in not adding it to the response to reduce the number of cases the client has to deal with in production.
If we added it in the response, then it would be ambiguous whether the client should care about it:
if we return it, it would be confusing for the client whether they have to enforce the policy they provided or the policy we return (they are supposed to be the same, but the fact we return it adds an ambiguity on how the client is supposed to behave). If we set the expectations that the sampling mode will be the one they passed with no exceptions, then there is only one possible behavior the client can take.

Adding it to the debug section of the response instead makes sense to me. I did not do that because it is exceptionally hard with the currently data structure to propagate debug information from query processors to parse_and_run query without adding it to the query object. I have in mind a separate data structure to add where query processors can add whatever they want and it will be added to the response and to query log, but I did not want to make a special case this for sampling rate, so that can wait in a separate PR.